precompile/contracts/warp/contract.go

```diff
@@ -0,0 +1,269 @@
+	// Return the packed output and the remaining gas
+	return packedOutput, remainingGas, nil
+}
+
+// UnpackSendWarpMessageInput attempts to unpack [input] as SendWarpMessageInput
+// assumes that [input] does not include selector (omits first 4 func signature bytes)
+func UnpackSendWarpMessageInput(input []byte) (SendWarpMessageInput, error) {
+	inputStruct := SendWarpMessageInput{}
+	err := WarpABI.UnpackInputIntoInterface(&inputStruct, "sendWarpMessage", input)
+
+	return inputStruct, err
+}
+
+// PackSendWarpMessage packs [inputStruct] of type SendWarpMessageInput into the appropriate arguments for sendWarpMessage.
+func PackSendWarpMessage(inputStruct SendWarpMessageInput) ([]byte, error) {
+	return WarpABI.Pack("sendWarpMessage", inputStruct.DestinationChainID, inputStruct.DestinationAddress, inputStruct.Payload)
+}
+
+func sendWarpMessage(accessibleState contract.AccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	if remainingGas, err = contract.DeductGas(suppliedGas, SendWarpMessageGasCost); err != nil {
+		return nil, 0, err
+	}
+	// This gas cost includes buffer room because it is based off of the total size of the input instead of the produced payload.
+	// This ensures that we charge gas before we unpack the variable sized input.
+	payloadGas, overflow := math.SafeMul(SendWarpMessageGasCostPerByte, uint64(len(input)))
+	if overflow {
+		return nil, 0, vmerrs.ErrOutOfGas
+	}
+	if remainingGas, err = contract.DeductGas(remainingGas, payloadGas); err != nil {
+		return nil, 0, err
+	}
+	if readOnly {
+		return nil, remainingGas, vmerrs.ErrWriteProtection
+	}
+	// unpack the arguemnts
+	inputStruct, err := UnpackSendWarpMessageInput(input)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	var (
+		sourceChainID      = accessibleState.GetSnowContext().ChainID
+		destinationChainID = inputStruct.DestinationChainID
+		sourceAddress      = caller.Hash()
+		destinationAddress = inputStruct.DestinationAddress
+		payload            = inputStruct.Payload
+	)
+
+	addressedPayload, err := warpPayload.NewAddressedPayload(
+		ids.ID(sourceAddress),
+		destinationAddress,
+		payload,
+	)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+	warpMessage, err := warp.NewUnsignedMessage(
+		sourceChainID,
+		destinationChainID,
+		addressedPayload.Bytes(),
+	)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	// Add a log to be handled if this action is finalized.
+	accessibleState.GetStateDB().AddLog(
+		ContractAddress,
+		[]common.Hash{
+			WarpABI.Events["SendWarpMessage"].ID,
+			destinationChainID,
+			destinationAddress,
```

Comment:
How did you decide on which topics to add to the log? And when receiving the log do we need to check that the ID/Addresses are the same? 

For example, parsing the data bytes back to warp unsigned message, then checking that topic[1] destination chain ID matches.

Reply:
Documented here: https://github.com/ava-labs/subnet-evm/blob/warp-e2e/precompile/contracts/warp/README.md#sendwarpmessage